Tracking/components/TrackdNdxDelphesBased.cpp

```diff
@@ -0,0 +1,174 @@
+
+DECLARE_COMPONENT(TrackdNdxDelphesBased)
+
+TrackdNdxDelphesBased::TrackdNdxDelphesBased(const std::string& name, ISvcLocator* svcLoc)
+    : Transformer(name, svcLoc,
+        {KeyValues("InputLinkCollection", {"TrackMCParticleLinks"}),
+         KeyValues("HeaderName", {"EventHeader"})},
+        {KeyValues("OutputCollection", {"MCParticles"})}) {}
+
+StatusCode TrackdNdxDelphesBased::initialize() {
+    m_uniqueIDSvc = service("UniqueIDGenSvc");
+    if (!m_uniqueIDSvc) {
+        error() << "Unable to locate the UniqueIDGenSvc" << endmsg;
+        return StatusCode::FAILURE;
+    }
+    m_geoSvc = service("GeoSvc");
+    if (!m_geoSvc) {
+        error() << "Unable to locate the GeoSvc" << endmsg;
+        return StatusCode::FAILURE;
+    }
+
+    // Initialise the delphes track util
+    m_delphesTrkUtil = new TrkUtil();
+    if (!m_delphesTrkUtil) {
+        error() << "Failed to create delphes TrkUtil instance" << endmsg;
+        return StatusCode::FAILURE;
+    }
+
+    // Load the geometry parameters from the XML file
+    double Rmin = m_geoSvc->getDetector()->constantAsDouble(m_Rmin_parameter_name.value());
+    double Rmax = m_geoSvc->getDetector()->constantAsDouble(m_Rmax_parameter_name.value());
+    double Zmax = m_geoSvc->getDetector()->constantAsDouble(m_Zmax_parameter_name.value());
+    double Zmin;
+    // For fwd-bwd symmetric detectors, use negative Zmax as Zmin
+    if (m_Zmin_parameter_name.value() == m_Zmax_parameter_name.value()) {
+        Zmin = -Zmax;
+    } else {
+        Zmin = m_geoSvc->getDetector()->constantAsDouble(m_Zmin_parameter_name.value());
+    }
+
+    debug() << "Geometry parameters:" << endmsg;
+    debug() << "Rmin: " << Rmin/dd4hep::cm << " cm" << endmsg;
+    debug() << "Rmax: " << Rmax/dd4hep::cm << " cm" << endmsg;
+    debug() << "Zmin: " << Zmin/dd4hep::cm << " cm" << endmsg;
+    debug() << "Zmax: " << Zmax/dd4hep::cm << " cm" << endmsg;
+
+    // dd4hep uses cm, while delphes (or more importantly the track parametrisation) uses mm, so we need to convert
+    m_delphesTrkUtil->SetDchBoundaries(Rmin/dd4hep::mm, Rmax/dd4hep::mm, Zmin/dd4hep::mm, Zmax/dd4hep::mm);
+
+    m_delphesTrkUtil->SetGasMix(m_GasSel.value());
+
+    // Make sure fill factor is between 0 and 1
+    if (m_fill_factor.value() < 0.0 || m_fill_factor.value() > 1.0) {
+        warning() << "Fill factor is not between 0 and 1, setting to 1.0" << endmsg;
+        m_fill_factor.set(1.0);
+    }
+
+    return StatusCode::SUCCESS;
+}
+
+edm4hep::RecDqdxCollection TrackdNdxDelphesBased::operator()(const edm4hep::TrackMCParticleLinkCollection& input,
+                                                            const edm4hep::EventHeaderCollection& header) const {
+    edm4hep::RecDqdxCollection outputCollection;
+
+    auto engine_seed = m_uniqueIDSvc->getUniqueID(header, this->name());
+    m_engine.seed(engine_seed);
+
+    // Dummy value if dN/dx calculation somehow fails
+    const double dummy_value = -999 * (1/dd4hep::m);
```

Comment:
Why do we need per meter now?

Reply:
Just asking why one dummy value is better than another. But I understand you multiply by the factor later and just want to get the known output number.